### ISSUE TITLE
Improve Preset Wizard

### ISSUE THREAD

**Issue by danielsuguimoto** [kool-dev/kool#312]
Today, the preset wizard is ok for the final user....but it is confusing for a contributor to change/add a preset.

- In `preset-config.yml` file, only questions under `compose` key can affect both `docker-compose.yml` and `kool.yml` files. Questions under `kool` key affect only `kool.yml`... we should make questions under `kool` affect both as well or create a new key, which affect both....and questions under `kool` and `compose` will affect only one file.
- We only create a file dynamically only if there's a question about it....but like I described before, a question about one file can affect another one....and it is not being considered today, so kool tries to get the default file if present...if not, the file isn't created

**Comment by fabriciojs** [comment]
The plan of action currently in mind:

- Change preset config structure to have a first-level array "questions".
- Each object in this array will be a question, with a default value, with its options and "actions" for each of them.
- An "action" could be anything that we wanna do/change in the project setup based on that choice, including but not limited to:
    - Merge a YAML file from a template onto the project.
    - Copy over a template file to the project.
    - Search/replace some content in a given file.
    - Others...
- Each of these actions would clearly specify their target files

This way we can accomplish in a generic way, any edits/changes we need to files in the current directory, giving more power and flexibility to presets creation.

**Comment by fabriciojs** [comment]
We should also add in support to comments on the parsed YAML files, as described on https://github.com/kool-dev/kool/issues/315

**Comment by fabriciojs** [comment]
- [ ] Create documentation once we have this (as per https://github.com/kool-dev/kool/issues/144)

**Comment by fabriciojs** [comment]
- [ ] As of https://github.com/kool-dev/kool/issues/227, we also want to implement that sort of automation within the Presets configuration scope.

**Comment by fabriciojs** [comment]
@kool-dev/core-members as per all mentioned above and in the related issues discussions, here is a new proposal for how a Preset configuration file should look like:

```yaml
# The name of the preset
name: 'Laravel App'

# Which language(s) are related to this preset
language: [ 'PHP' ]

# Create defines the workflow for creating a new Project where this preset can then be installed
create:
  steps:
    - name: Creating new Laravel Application
      apply:
        - scripts:
           - docker pull -q kooldev/php:7.4
           - kool docker kooldev/php:7.4 composer create-project --no-install --no-scripts --prefer-dist laravel/laravel $CREATE_DIRECTORY

# Preset defines the workflow for installing this preset in the current working directory
preset:
  steps:
    - name: 'Copy basic config files'
      apply:
        - copy: docker-compose.yml
          as: docker-compose.yml
        - copy: kool.yml
          as: kool.yml

    # Defines which app service to use (PHP version)
    - ask: 'Which app service do you want to use'
      default: 'PHP 8.0'
      options:
        - name: 'PHP 8.0'
          apply:
            - snippet: php-8
        - name: 'PHP 7.4'
          apply:
            - snippet: php-7.4

    # Defines which database service to use (DB)
    - ask: 'Which database service do you want to use'
      default: 'MySQL 8.0'
      options:
        - name: 'MySQL 8.0'
          apply:
            - snippet: mysql-8
        - name: 'MySQL 5.7'
          apply:
            - snippet: mysql-5.7
        - name: 'MariaDB 10.5'
          apply:
            - snippet: maria-10.5
        - name: 'PostgreSQL 13.0'
          apply:
            - snippet: postgresql-13
        - name: 'None - do not use a database'

    # Defines which cache service to use (in-memory key/value storage)
    - ask: 'Which cache service do you want to use'
      default: 'Redis 6.0'
      options:
        - name: 'Redis 6.0'
          apply:
            - snippet: redis-6
        - name: 'Memcached 1.6'
          apply:
            - snippet: memcached-1.6
        - name: 'None - do not use a key/value storage'
      options:
        - name: 'Redis 6.0'
          apply:
            - snippet: redis-6
        - name: 'Memcached 1.6'
          apply:
            - snippet: memcached-1.6
        - name: 'None - do not use a key/value storage'

    # Defines which Javascript package manager to use
    - ask: 'Which Javascript package manager do you want to use'
      default: 'NPM'
      options:
        - name: 'NPM'
          apply:
            - snippet: npm
        - name: 'Yarn'
          apply:
            - snippet: yarn
        - name: 'None'

```

Please don't continue reading before you try to digest/understand, at first sight, what each of the definitions above means and how they work. Once you read the explanations below, we lose the "clear mind, the first impression".


---


<details>
  <summary>Configuration scheme</summary>

- `steps` it's always a list of Steps. Each step is an object with either a simple name/label, or a question (`ask` action), and each one must have a list of `actions` to perform.
  
- `apply` it's always a list/array of **Actions**. Each action can be one of the pre-defined actions we have: 
    - `ask:` - is an action that prompts for some input and then applies a given set of *actions*.  
    - `copy:` - copy over some file.
    -  `replace:` - performs a text replacement with Regex. Meant for automatic changes of `.env` files for example.
    - `merge:` - merge files/partials - meant to be used for composing YAML files like `docker-compose.yml` or ` kool.yml`.
    - `snippet` - it's a call to another set of **steps** defined in a *snippet*... this is meant for reuse, it's recursive, so it can include any other action/snippets... the difference from a preset and a snippet is for our understanding only, in practice they can behave the same. But presets will be larger/for a full environment, and snippets handle pieces only used for composition. 


</details>

**Comment by bessa3301** [comment]
@fabriciojs is the cache options meant to be there twice?
```
 # Defines which cache service to use (in-memory key/value storage)
    - ask: 'Which cache service do you want to use'
      default: 'Redis 6.0'
      options:
        - name: 'Redis 6.0'
          apply:
            - snippet: redis-6
        - name: 'Memcached 1.6'
          apply:
            - snippet: memcached-1.6
        - name: 'None - do not use a key/value storage'
      # 2nd time ?
      options:
        - name: 'Redis 6.0'
          apply:
            - snippet: redis-6
        - name: 'Memcached 1.6'
          apply:
            - snippet: memcached-1.6
        - name: 'None - do not use a key/value storage'
```

**Comment by fabriciojs** [comment]
@bessa3301 no, it should be only one!

**Comment by fabriciojs** [comment]
@kool-dev/core-members as per initial discussions, the path we are aiming at right now:


### Generic Kool Automation Feature

- `steps`: a list of steps given automation has. Each step can be either a *simple step* (with a `name` and `actions` list) or a question (with `ask`).
- `actions`: a list of actions to be performed. Each action can be of one of the predefined types, with its own arguments.
- `ask`: a prompt to the user to pick one of the available options. Useful for customization as per user-specific choices. Must contain a `default` that refers to one of the options; `options` is the list of options available, where each option should have a `name` and an optional `actions` list, to whatever actions should be executed when that option is selected.

### Possible Automations within Kool
- Current: Preset Create automation
        $ kool create PRESET
- Current: Preset Install automation
        $ kool preset PRESET
- Proposal/future: Snippet Install automation - this would be a single service. i.e elastisearch, mongodb, redis, koolpdf, etc.
        $ kool add SNIPPET

### Actions

Each action within the `actions` list key should be one of the predefined types:

- `copy` - copies a file from internal storage or in the local directory.
- `merge` - merges YAML files.
- `replace` - replaces strings, using or not REGEX.
- `scripts` - run one or more arbitrary commands (shell).

**Comment by wesleyhf** [comment]
@fabriciojs I like the structure and the idea, it is pretty useful. Just one question regarding to `language` key.

```yaml
# Which language(s) are related to this preset
language: [ 'PHP' ]
```

What is the main purpose of this key?

If I understood it correctly this key can be replaced to `tags`, or something like that.

**Comment by danielsuguimoto** [comment]
> @fabriciojs I like the structure and the idea, it is pretty useful. Just one question regarding to `language` key.
> 
> ```yaml
> # Which language(s) are related to this preset
> language: [ 'PHP' ]
> ```
> 
> What is the main purpose of this key?
> 
> If I understood it correctly this key can be replaced to `tags`, or something like that.

First question on preset command is the language
![image](https://user-images.githubusercontent.com/25229545/126151760-d1a3fba6-cf00-46a2-850c-c1bbb11675a0.png)

**Comment by fabriciojs** [comment]
Jun's answer is precisely correct. Although, I can imagine we stretching the meaning of this to be `tags`, or a more generic grouping for presets... possibly we could have something other than a language per say being used to branch the choices on `kool preset`?

**Comment by fabriciojs** [comment]
This has been implemented within our [latest release 1.15.0](https://github.com/kool-dev/kool/releases/tag/1.15.0).
